### PR TITLE
Ro stew/sync project tags with field and label files

### PR DESCRIPTION
**Pull request by stew-ro**

### CODE REVIEW

File: src/redux/actions/projectActions.ts [hunk 1]
```diff
@@ -88,6 +90,18 @@ export function saveProject(project: IProject)
...
 }
 
+export function updateProjectTagsFromFiles(project: IProject)
+    : (dispatch: Dispatch, getState: () => IApplicationState) => Promise<void> {
+    return async (dispatch: Dispatch, getState: () => IApplicationState) => {
+        const projectCopy = Object.assign({}, project);
+        const projectService = new ProjectService();
+        const storageProvider = StorageProviderFactory.createFromConnection(project.sourceConnection);
+        await projectService.getTagsFromPreExistingFieldFile(projectCopy, storageProvider);
+        await projectService.getTagsFromPreExistingLabelFiles(projectCopy, storageProvider);
```
Comment (kunzms) [inline src/redux/actions/projectActions.ts]: Do we still need label files?

Comment (stew-ro) [reply src/redux/actions/projectActions.ts]: I modified getTagsFromPreExistingLabelFiles to only add project tags that are missing from the label file. Before, it was overriding the current project tags from the labels.

Comment (kunzms) [reply src/redux/actions/projectActions.ts]: Sound good.

File: src/services/projectService.ts [hunk 2]
```diff
@@ -197,7 +199,8 @@ export default class ProjectService implements IProjectService {
                 } as ITag);
             });
             if (project.tags) {
-                project.tags = patch(tags, project.tags, "name", ["color"]);
+                const newTags = tags.filter((fileTag) => !project.tags.find((tag) => fileTag.name === tag.name ));
+                project.tags = [...project.tags, ...newTags];
```
Comment (kunzms) [inline src/services/projectService.ts]: put this logic into a method so we can remove duplicate codes.

File: src/services/projectService.ts [hunk 3]
```diff
@@ -230,6 +233,8 @@ export default class ProjectService implements IProjectService {
             });
             if (project.tags) {
                 project.tags = patch(project.tags, tags, "name", ["type", "format"]);
+                const newTags = tags.filter((fileTag) => !project.tags.find((tag) => fileTag.name === tag.name ));
```
Comment (kunzms) [inline src/services/projectService.ts]: Will we assign color to new tag property?

Comment (stew-ro) [reply src/services/projectService.ts]: yes, in the setColorsForUpdatedTags function, a project's existing tag's color wont change and all new tags will get the next color to be used from tagColors.json

File: src/redux/actions/projectActions.ts [hunk 4]
```diff
@@ -88,6 +90,20 @@ export function saveProject(project: IProject)
...
+export function updateProjectTagsFromFiles(project: IProject)
+    : (dispatch: Dispatch, getState: () => IApplicationState) => Promise<void> {
+    return async (dispatch: Dispatch, getState: () => IApplicationState) => {
+        const updatedProject = Object.assign({}, project);
+        updatedProject.tags = [];
+        const projectService = new ProjectService();
+        const storageProvider = StorageProviderFactory.createFromConnection(project.sourceConnection);
+        await projectService.getTagsFromPreExistingFieldFile(updatedProject, storageProvider);
+        await projectService.getTagsFromPreExistingLabelFiles(updatedProject, storageProvider);
+        await projectService.setColorsForUpdatedTags(project, updatedProject);
```
Comment (kunzms) [inline src/redux/actions/projectActions.ts]: add a new public updateProjectFromFiles(project): IProject to projectService
and call these function internally and make them private.

**Pull request merged**